include/pybind11/cast.h

```diff
@@ -434,11 +438,19 @@ struct string_caster {
     cast(const StringType &src, return_value_policy /* policy */, handle /* parent */) {
         const char *buffer = reinterpret_cast<const char *>(src.data());
         auto nbytes = ssize_t(src.size() * sizeof(CharT));
-        handle s = decode_utfN(buffer, nbytes);
-        if (!s) {
-            throw error_already_set();
+        if (std::is_same_v<CharT, uint8_t>
+#if defined(PYBIND11_CPP17)
+            or std::is_same_v<CharT, std::byte>
```

Comment:
You also may need to wrap this in special MACRO to prevent it MSVC from complaining.

Reply:
regarding that flag, what's the threshold for introducing a new PYBIND11_xxx flag (such as `PYBIND11_HAS_U8STRING`, `PYBIND11_HAS_STRING_VIEW`) to hide a check for presence of `std::byte`?
I went for that simple check since std::byte is introduced with c++17 but you never know what c++ partial-support compiler versions their could be out there.